consensus/hotstuff/signature/signature_aggregator.go

```diff
@@ -1,33 +1,57 @@
 package signature
 
 import (
+	"fmt"
+
 	"github.com/onflow/flow-go/consensus/hotstuff"
 	"github.com/onflow/flow-go/crypto"
 	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module"
 )
 
-// SignatureAggregatorImpl aggregates the signatures
+// SignatureAggregatorImpl implements hotstuff.SignatureAggregator interface
+// It aggregates signatures and tracks aggregated weight. This aggregator can be used both for staking signatures and
+// threshold signatures aggregation. It will be included into CombinedAggregator to track staking and threshold signatures.
+// CombinedAggregator will handle concurrency access to this primitive.
+// This structure is NOT concurrency safe.
 type SignatureAggregatorImpl struct {
+	sigCollector flow.SignatureCollector
+	aggregator   module.AggregatingSigner
+	totalWeight  uint64
 }
 
 var _ hotstuff.SignatureAggregator = &SignatureAggregatorImpl{}
 
+func NewSignatureAggregator(aggregator module.AggregatingSigner) *SignatureAggregatorImpl {
+	return &SignatureAggregatorImpl{
+		aggregator:   aggregator,
+		sigCollector: flow.NewSignatureCollector(),
+		totalWeight:  0,
+	}
+}
+
 // TrustedAdd adds an already verified signature, and look up the weight for the given signer,
 // and add it to the total weight, and returns the total weight that have been collected.
 // return (1000, nil) means the signature has been added, and 1000 weight has been collected in total.
 // return (1000, nil) means the signature is a duplication and 1000 weight has been collected in total.
-func (s *SignatureAggregatorImpl) TrustedAdd(signerID flow.Identifier, sig crypto.Signature) (totalWeight uint64, exception error) {
-	panic("to be implemented")
+func (s *SignatureAggregatorImpl) TrustedAdd(signerID flow.Identifier, weight uint64, sig crypto.Signature) (totalWeight uint64, exception error) {
+	if hasSigned := s.sigCollector.HasSigned(signerID); hasSigned {
+		return 0, fmt.Errorf("signature for (%x) was already collected", signerID)
+	}
+
+	s.sigCollector.Add(signerID, sig)
+	s.totalWeight += weight
+	return s.totalWeight, nil
```

Comment:
The error is for exceptions, but this case is not an exception. 

Please refer to [this comment](https://github.com/onflow/flow-go/pull/1164/files#diff-faa591ccd4ac38232acffa4b2fcf61532c45e459d3a99b2e203a2489269f7a4cR36)

It's better for `SigCollector.Add` to return an `added bool`, depending on the result we can add weight or not.
```suggestion
	added := s.sigCollector.Add(signerID, sig)
	if added {
		s.totalWeight += weight
	}
	return s.totalWeight, nil
```